src/js/libcs/Json.js

```diff
@@ -0,0 +1,223 @@
+                    erg = erg + ', ';
+                }
+
+            }
+            return erg + "]";
+        }
+
+        return "\"" + String(niceprint(a)) + "\"";
+    }
+
+    return niceprint(a);
+};
+
+Json.Atomniceprint = function(el) {
+    return "\{" + el.value.key + ":" + niceprint(el.value.value) + "\}";
+};
+
+Json.niceprint = function(el, modifs, options) {
+    if (!options) {
+        options = {};
+        options.printValidJSON = false;
+        options.printedWarning = false;
+        // track depth
+        options.visitedMap = {};
+        options.visitedMap.level = 0;
+        options.visitedMap.maxLevel = 1000;
+        options.visitedMap.maxElVisit = 5000;
+
+        if (modifs) {
+            if (modifs.JSON) {
+                let jmodif = evaluate(modifs.JSON);
+                if (jmodif.ctype === "boolean") options.printValidJSON = jmodif.value;
+            }
+            if (modifs.maxDepth) {
+                let depth = evaluate(modifs.maxDepth);
+                if (depth.ctype === "number") options.visitedMap.maxLevel = depth.value.real;
+            }
+        }
+    }
+
+    var visitedMap = options.visitedMap;
+    // track a new recursive call
+    visitedMap.newLevel = true;
+    visitedMap.level += 1;
+
+    var keys = Object.keys(el.value).sort();
+    var jsonString = "{" + keys.map(function(key) {
+        // update visitedMap
+        let elValKey = el.value[key];
+        if (!visitedMap[elValKey]) {
+            visitedMap[elValKey] = 1;
+        } else {
+            if (visitedMap[elValKey] > visitedMap.maxElVisit || visitedMap.level > visitedMap.maxLevel) {
+                //console.log([visitedMap[elValKey], visitedMap.level]);
+                if (!options.printedWarning) {
+                    console.log("Warning: We visited a key-value pair very often or encountered a very deeply nested dictionary. Dictionary is probably cyclic. Output will be probably incomplete.");
+                    options.printedWarning = true;
+                }
+                if (options.printValidJSON) {
+                    return "\"" + key + "\"" + ":" + '"..."';
+                } else {
+                    return key + ":" + '...';
+                }
+            }
+            // update only once a recursive call
+            if (visitedMap.newLevel) {
+                visitedMap[elValKey] += 1;
+                // update only once each function call
+                visitedMap.newLevel = false;
+            }
+        }
+        if (options.printValidJSON) {
+            return "\"" + key + "\"" + ":" + Json._helper.niceprint(elValKey, modifs, options);
+        } else {
+            return key + ":" + Json._helper.niceprint(elValKey, modifs, options);
+        }
+    }).join(", ") + "}";
+
+
+    // pretty print 
+    // to be valid JSON we need to replace single with double quotes
+    if (options.printValidJSON) {
+        jsonString = jsonString.replace(/'/g, '"');
```

Comment:
I am not sure on that. `{"a'":1}` should be valid JSON. However, we should escape double quotes and newlines within the keys/values. They are hard to enter in CindyJS:
```
evilstr1 = unicode("0022");
evilstr2 = "
"
json = {"a":evilstr1, "b":evilstr2};
print(json); 
```
This gives
```
{a:", b:
}
```
valid JSON could be something like `{"a":"\"", "b":"\n"}`. Should we use the `encodeURI()` function for key in line 197 and all printed strings?

Reply:
update: better use the `JSON.stringify`-method. It does exactly what we need.
We can apply it either to the plain strings.